File: src/test/java/br/com/unopay/api/AuthServerApplicationTests.java
package br.com.unopay.api;


import br.com.unopay.api.uaa.model.UserDetail;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.base.Throwables;
import com.jayway.jsonpath.JsonPath;
import org.junit.Before;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.crypto.codec.Base64;
import org.springframework.security.web.FilterChainProxy;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.test.context.web.WebAppConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import java.io.UnsupportedEncodingException;

import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public abstract class AuthServerApplicationTests  extends  UnopayApiApplicationTests{

    protected ResultActions clientCredentials() throws Exception {
        return this.mvc.perform(post("/oauth/token")
                .contentType(
                        MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "client")
                .param("client_secret", "secret"));
    }


    protected ResultActions passwordFlow(String username, String password) throws Exception {
        return this.mvc.perform(post("/oauth/token")
                .contentType(
                        MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "password")
                .param("client_id", "client")
                .param("client_secret", "secret")
                .param("username", username)
                .param("password", password));
    }


    protected ResultActions uaaManagerClientCredentials() throws Exception {
        return this.mvc.perform(post("/oauth/token")
                .contentType(
                        MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "manager")
                .param("client_secret", "secret"));
    }

    protected ResultActions wrongClientCredentials() throws Exception {
        return this.mvc.perform(post("/oauth/token")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "wrong")
                .param("client_secret", "wrong"));
    }

    protected String getAuthorizationHeader(String clientId, String clientSecret) {
        String creds = String.format("%s:%s", clientId, clientSecret);
        try {
            return "Basic " + new String(Base64.encode(creds.getBytes("UTF-8")));
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException("Could not convert String");
        }
    }

    protected String toJson(UserDetail user) {
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            return objectMapper.writeValueAsString(user);
        } catch (JsonProcessingException e) {
            Throwables.propagate(e);
            return null;
        }
    }

    protected String clientCredentialsAccessToken() throws Exception {
        MvcResult result = clientCredentials()
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.access_token", is(notNullValue())))
                .andReturn();

        return JsonPath.read(
                result.getResponse().getContentAsString(),
                "$.access_token");
    }

}

File: src/test/scala/br/com/unopay/api/AuthServerApplicationTests.scala
package br.com.unopay.api

import java.io.UnsupportedEncodingException

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.jayway.jsonpath.JsonPath
import org.hamcrest.core.Is._
import org.hamcrest.core.IsNull._
import org.springframework.http.MediaType
import org.springframework.security.crypto.codec.Base64
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders._
import org.springframework.test.web.servlet.result.MockMvcResultMatchers._
import org.springframework.test.web.servlet.{MvcResult, ResultActions}

trait AuthServerApplicationTests extends UnopayApiScalaApplicationTest {

    protected def getClientAccessToken(): String = {
        val result = clientCredentials()
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.access_token", is(notNullValue())))
                .andReturn()
        getAccessToken(result)
    }

    protected def getUserAccessToken(user: String = "test@test.com",  pwd: String = "test"): String = {
        val result = passwordFlow(user, pwd)
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.access_token", is(notNullValue())))
                .andReturn()
        getAccessToken(result)
    }


    protected def clientCredentials(): ResultActions = {
        this.mvc.perform(post("/oauth/token")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "client")
                .param("client_secret", "secret"))
    }


    protected def passwordFlow(username: String,  password: String): ResultActions = {
        return this.mvc.perform(post("/oauth/token")
                .contentType(
                MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "password")
                .param("client_id", "client")
                .param("client_secret", "secret")
                .param("username", username)
                .param("password", password))
    }


    protected def uaaManagerClientCredentials(): ResultActions ={
        return this.mvc.perform(post("/oauth/token")
                .contentType(
                MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "manager")
                .param("client_secret", "secret"))
    }

    protected def  wrongClientCredentials(): ResultActions =  {
        this.mvc.perform(post("/oauth/token")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .param("grant_type", "client_credentials")
                .param("client_id", "wrong")
                .param("client_secret", "wrong"))
    }

    protected def getAuthorizationHeader(clientId: String, clientSecret: String): String=  {
        val creds = String.format("%s:%s", clientId, clientSecret)
        try {
            "Basic " + new String(Base64.encode(creds.getBytes("UTF-8")))
        } catch {
            case e: UnsupportedEncodingException => throw new IllegalStateException("Could not convert String")
        }
    }

    override def toJson(obj: Object): String = {
        try {
            val objectMapper = new ObjectMapper()
            objectMapper.writeValueAsString(obj)
        } catch {
            case e: JsonProcessingException => throw  e
        }
    }

    protected def toJsonFromView(obj: Object,  view: Class[_]): String = {
        try {
            val objectMapper = new ObjectMapper()
            return objectMapper.writerWithView(view).writeValueAsString(obj)
        } catch {
            case e: JsonProcessingException => throw  e
        }
    }

    protected def getLocationHeader(mvcResult: MvcResult): String = {
        mvcResult.getResponse.getHeader("Location")
    }


    protected def  getUAAManagerAccessToken(): String =  {
        val result = uaaManagerClientCredentials()
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.access_token", is(notNullValue())))
                .andReturn()
        getAccessToken(result)
    }

    protected def clientCredentialsAccessToken(): String = {
        val result = clientCredentials()
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.access_token", is(notNullValue())))
                .andReturn()

        JsonPath.read[String](result.getResponse.getContentAsString, "$.access_token")
    }

    protected def getAccessToken(result: MvcResult): String = {
        JsonPath.read[String](
                result.getResponse.getContentAsString,
            "$.access_token")
    }

}
